ci/upload_metrics_to_s3.py

```diff
@@ -0,0 +1,329 @@
+        if self.environment == "production":
+            self.domain_prefix = DOMAIN_PREFIX +"prd-"
+        if self.environment == "staging":
+            self.domain_prefix = DOMAIN_PREFIX +"stg-"
+        if self.environment == "development":
+            self.domain_prefix = DOMAIN_PREFIX +"dev-"
+        self.is_daily = False
+
+
+    # some metrics need a specific instance called for them
+    def get_instance_ids_for_domain(self,namespace,domain,metric):
+
+        response = cloudwatch_client.list_metrics(
+            Namespace=namespace,
+            Dimensions = [{"Name": "DomainName", "Value": domain}],
+            MetricName=metric["name"],
+
+        )
+        instance_ids = []
+        for metric in response["Metrics"]:
+            for dim in metric["Dimensions"]:
+                if dim["Name"] == "NodeId":
+                    instance_ids.append(dim["Value"])
+        return instance_ids
+
+    def get_metric_logs(self, start, end, namespace, metric, dimensions,period, statistic,tags, instance=None):
+        response= cloudwatch_client.get_metric_statistics(
+            Namespace=namespace,
+            MetricName=metric["name"],
+            Dimensions=dimensions,
+            StartTime=start,
+            EndTime=end,
+            Period=period,
+            Statistics=statistic,
+            Unit=metric["unit"]
+        )
+        datapoints = response.get("Datapoints",[])
+
+        if not datapoints:
+            return []
+
+        metric_events = []
+        for dp in datapoints:
+            dp["Tags"] = tags
+            dp["MetricName"] = metric["name"]
+            dp["InstanceName"] = instance
+            dp["Time"] = dp["Timestamp"].isoformat()
+            dp.pop("Timestamp", None)
+            metric_events.append(dp)
+        return metric_events
+
+    def get_cg_domains(self):
+        domain_list = es_client.list_domain_names()
+        domain_names = [d['DomainName'] for d in domain_list['DomainNames']]
+        matching_domains = [d for d in domain_names if d.startswith(self.domain_prefix)]
+        return matching_domains
+
+    # Upload a batch of metric events to S3 as a single object
+    def put_metric_events_to_s3(self, object_name, metric_events):
+        body = "\n".join(
+            [
+                json.dumps(metric_event)
+                for metric_event in metric_events
+            ]
+        )
+        s3_client.put_object(
+            Bucket=self.bucket_name,
+            Key=object_name,
+            Body=body,
+            ContentType="text/plain",
+            ServerSideEncryption='AES256'
+        )
+
+    def update_latest_stamp_in_s3(self, latest_timestamp):
+        data = latest_timestamp
+        s3_client.put_object(
+            Bucket=self.bucket_name,
+            Key=timestamp_key,
+            Body=data,
+            ServerSideEncryption='AES256'
+        )
+
```

Comment:
If these two functions are exactly the same except for the `Key` value, I'd prefer to make the key an argument to a single function and use it for both cases